src/order/semiconj_Sup.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2020 Yury G. Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Yury G. Kudryashov
+-/
+import order.conditionally_complete_lattice
+import logic.function.conjugate
+import order.ord_continuous
+import data.equiv.mul_add
+
+/-!
+# Semiconjugate by `Sup`
+
+In this file we prove two facts about semiconjugate (families of) functions.
+
+First, if an order isomorphism `fa : α → α` is semiconjugate to an order embedding
+`fb : β → β` by `g : α → β`, then `fb` is semiconjugate to `fa` by `y ↦ Sup {x | g x ≤ y}`.
+In particular, the relation “semiconjugate” is an equivalence relation on order
+isomorphisms `α → α`.
+
+Second, consider two actions `f₁ f₂ : G → α → α` of a group on a complete lattice by order
+isomorphisms. Then the map `x ↦ ⨆ g : G, (f₁ g)⁻¹ (f₂ g x)` semiconjugates each `f₁ g'` to `f₂ g'`.
+In the case of a conditionally complete lattice, a similar statement holds true under an
+additional assumption that each set `{(f₁ g)⁻¹ (f₂ g x) | g : G}` is bounded above.
+
+The lemmas come from [Étienne Ghys, Groupes d'homeomorphismes du cercle et cohomologie
+bornee][ghys87:groupes], Proposition 2.1 and 5.4 respectively. In the paper they are formulated for
+homeomorphisms of the circle, so in order to apply results from this file one has to lift these
+homeomorphisms to the real line first.
+-/
+
+variables {α : Type*} {β : Type*}
+
+open set
+
+/-- We say that `g : β → α` is an order right adjoint function for `f : α → β` if it sends each `y`
+to a least upper boundary for `{x | f x ≤ y}`. If `α` is a partial order, and `f : α → β` has
+a right adjoint, then this right adjoint is unique. -/
+def is_order_right_adjoint [preorder α] [preorder β] (f : α → β) (g : β → α) :=
+∀ y, is_lub {x | f x ≤ y} (g y)
+
+lemma is_order_right_adjoint.unique [partial_order α] [preorder β] {f : α → β} {g₁ g₂ : β → α}
+  (h₁ : is_order_right_adjoint f g₁) (h₂ : is_order_right_adjoint f g₂) :
+  g₁ = g₂ :=
+funext $ λ y, (h₁ y).unique (h₂ y)
+
+lemma is_order_right_adjoint.right_mono [preorder α] [preorder β] {f : α → β} {g : β → α}
+  (h : is_order_right_adjoint f g) :
+  monotone g :=
+λ y₁ y₂ hy, (h y₁).mono (h y₂) $ λ x hx, le_trans hx hy
+
+namespace function
+
+lemma semiconj.symm_adjoint [partial_order α] [preorder β]
```

Comment:
This lemma (and the next ones) deserve docstrings, because the statement is not obvious. Maybe also mention the reference to Etienne's paper.